File: scripts/download-gutenberg.py
#!/usr/bin/env python3
"""
Download Gutenberg BookCorpus dataset from Hugging Face
and convert to SQLite for the literary-compiler MCP parser.

Dataset: incredible45/Gutenberg-BookCorpus-Cleaned-Data-English
Format: parquet -> SQLite (gutenberg.db)
"""

import os
import sys
import time
import sqlite3
import tempfile
import shutil
from pathlib import Path

# ─── Config ──────────────────────────────────────────────────────────────────

DATASET_ID = "incredible45/Gutenberg-BookCorpus-Cleaned-Data-English"
DATA_DIR = Path(__file__).parent.parent / "data" / "gutenberg"
DB_PATH = DATA_DIR / "gutenberg.db"
BATCH_SIZE = 1000

# ─── Main ────────────────────────────────────────────────────────────────────

def ensure_deps():
    """Check and install required Python packages."""
    try:
        import pyarrow.parquet
        import huggingface_hub
    except ImportError:
        print("Installing dependencies...")
        os.system(f"{sys.executable} -m pip install -q pyarrow huggingface_hub datasets")


def download_parquet_files(cache_dir: str) -> list[str]:
    """Download parquet files from Hugging Face."""
    from huggingface_hub import snapshot_download

    print(f"Downloading dataset {DATASET_ID}...")
    start = time.time()

    path = snapshot_download(
        repo_id=DATASET_ID,
        repo_type="dataset",
        local_dir=cache_dir,
        resume_download=True,
    )

    elapsed = time.time() - start
    print(f"Downloaded in {elapsed:.1f}s -> {path}")

    # Find all parquet files
    parquet_files = []
    for root, dirs, files in os.walk(path):
        for f in files:
            if f.endswith(".parquet"):
                parquet_files.append(os.path.join(root, f))

    print(f"Found {len(parquet_files)} parquet file(s)")
    return parquet_files


def create_sqlite_db(db_path: Path) -> sqlite3.Connection:
    """Create SQLite database with schema matching GutenbergParser expectations."""
    db_path.parent.mkdir(parents=True, exist_ok=True)

    conn = sqlite3.connect(str(db_path))
    conn.execute("PRAGMA journal_mode=WAL")
    conn.execute("PRAGMA synchronous=NORMAL")
    conn.execute("PRAGMA cache_size=-64000")  # 64MB cache

    conn.execute("""
        CREATE TABLE IF NOT EXISTS gutenberg (
            etextno INTEGER PRIMARY KEY,
            book_title TEXT,
            author TEXT,
            issued TEXT,
            context TEXT
        )
    """)

    conn.execute("""
        CREATE INDEX IF NOT EXISTS idx_gutenberg_author ON gutenberg(author)
    """)

    conn.commit()
    return conn


def parquet_to_sqlite(parquet_files: list[str], conn: sqlite3.Connection):
    """Stream parquet files into SQLite."""
    import pyarrow.parquet as pq

    total_rows = 0
    start = time.time()

    for pf in parquet_files:
        print(f"Processing {os.path.basename(pf)}...")
        pf_start = time.time()

        # Read in batches to avoid memory issues with large files
        parquet_file = pq.ParquetFile(pf)
        batch_count = 0

        for batch in parquet_file.iter_batches(batch_size=BATCH_SIZE):
            table = batch.to_pydict()
            rows = list(zip(
                table.get("etextno", []),
                table.get("book_title", []),
                table.get("author", []),
                table.get("issued", []),
                table.get("context", []),
            ))

            conn.executemany(
                "INSERT OR IGNORE INTO gutenberg (etextno, book_title, author, issued, context) VALUES (?, ?, ?, ?, ?)",
                rows,
            )

            batch_count += len(rows)
            total_rows += len(rows)

        conn.commit()
        elapsed = time.time() - pf_start
        print(f"  -> {batch_count} rows in {elapsed:.1f}s")

    elapsed = time.time() - start
    print(f"\nTotal: {total_rows} rows in {elapsed:.1f}s")
    return total_rows


def main():
    ensure_deps()

    # Use a temp dir for download, then move to final location
    with tempfile.TemporaryDirectory(prefix="gutenberg_") as tmpdir:
        parquet_files = download_parquet_files(tmpdir)

        if not parquet_files:
            print("ERROR: No parquet files found!")
            sys.exit(1)

        # Create database
        if DB_PATH.exists():
            print(f"Removing existing database: {DB_PATH}")
            DB_PATH.unlink()
            # Also remove WAL/SHM
            for ext in ["-wal", "-shm"]:
                p = DB_PATH.with_suffix(DB_PATH.suffix + ext)
                if p.exists():
                    p.unlink()

        conn = create_sqlite_db(DB_PATH)

        try:
            total = parquet_to_sqlite(parquet_files, conn)

            # Final stats
            count = conn.execute("SELECT COUNT(*) FROM gutenberg").fetchone()[0]
            size_mb = DB_PATH.stat().st_size / (1024 * 1024)
            print(f"\nDatabase ready: {DB_PATH}")
            print(f"  Rows: {count}")
            print(f"  Size: {size_mb:.1f} MB")
            print(f"\nSample entries:")
            for row in conn.execute(
                "SELECT etextno, book_title, author, LENGTH(context) as text_len FROM gutenberg LIMIT 5"
            ):
                print(f"  [{row[0]}] {row[1][:60]} by {row[2]} ({row[3]:,} chars)")
        finally:
            conn.close()


if __name__ == "__main__":
    main()

File: scripts/download-gutenberg.ts
#!/usr/bin/env bun
/**
 * Download Gutenberg BookCorpus from HuggingFace → SQLite.
 *
 * Uses direct HTTP (bypasses xet/2FA) — public repos only.
 * Outputs JSON progress lines to stdout for SSE integration.
 *
 * Usage: bun scripts/download-gutenberg.ts [--dir path] [--db path] [--workers N]
 */

import { readParquet } from "parquet-wasm";
import { tableFromIPC, type Table } from "apache-arrow";
import { Database } from "bun:sqlite";
import { mkdirSync, existsSync, statSync, readdirSync, readFileSync } from "node:fs";
import { join, basename } from "node:path";

// ── Args ──────────────────────────────────────────────────────

const args = process.argv.slice(2);
function flag(name: string, fallback: string): string {
  const i = args.indexOf(name);
  return i !== -1 && args[i + 1] ? args[i + 1] : fallback;
}
const PARQUET_DIR = flag("--dir", "data/mcp/gutenberg-bookcorpus");
const DB_PATH = flag("--db", "data/mcp/gutenberg-bookcorpus.db");
const MAX_WORKERS = parseInt(flag("--workers", "4"), 10);
const BATCH_SIZE = 5000;
const RETRY = 3;

const DATASET_ID = "incredible45/Gutenberg-BookCorpus-Cleaned-Data-English";

// ── Progress: JSON lines to stdout ────────────────────────────

interface ProgressMsg {
  phase: string;
  pct: number;
  message: string;
}

function emit(phase: string, pct: number, message: string) {
  const msg: ProgressMsg = { phase, pct: Math.round(pct), message };
  console.log(JSON.stringify(msg));
}

// ── Step 1: List parquet files from HF API ────────────────────

async function listParquetFiles(): Promise<string[]> {
  const url = `https://huggingface.co/api/datasets/${DATASET_ID}/tree/main/data`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`HF API error: ${res.status} ${res.statusText}`);
  const files = (await res.json()) as Array<{ path: string }>;
  return files.filter((f) => f.path.endsWith(".parquet")).map((f) => f.path);
}

// ── Step 2: Download with concurrency ─────────────────────────

async function downloadOne(filename: string, dest: string): Promise<void> {
  const url = `https://huggingface.co/datasets/${DATASET_ID}/resolve/main/${filename}`;

  for (let attempt = 0; attempt < RETRY; attempt++) {
    try {
      const res = await fetch(url);
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const buf = await res.arrayBuffer();
      await Bun.write(dest, buf);
      return;
    } catch (err) {
      if (attempt < RETRY - 1) {
        await Bun.sleep(1000 * 2 ** attempt);
      } else {
        throw err;
      }
    }
  }
}

async function downloadAll(files: string[]): Promise<number> {
  mkdirSync(PARQUET_DIR, { recursive: true });

  const tasks: Array<{ name: string; dest: string }> = [];
  let skipped = 0;

  for (const fname of files) {
    const dest = join(PARQUET_DIR, basename(fname));
    if (existsSync(dest) && statSync(dest).size > 10000) {
      skipped++;
      continue;
    }
    tasks.push({ name: fname, dest });
  }

  if (tasks.length === 0) {
    emit("download", 100, `All ${skipped} files already cached`);
    return skipped;
  }

  emit("download", 0, `${skipped} cached, ${tasks.length} to download (${MAX_WORKERS} workers)`);

  let done = 0;
  const total = tasks.length;

  const executing = new Set<Promise<void>>();

  for (const task of tasks) {
    const p = downloadOne(task.name, task.dest).then(() => {
      executing.delete(p);
      done++;
      const mb = statSync(task.dest).size / 1048576;
      emit("download", (done / total) * 100, `${done}/${total} ${basename(task.name)} (${mb.toFixed(0)}MB)`);
    });
    executing.add(p);

    if (executing.size >= MAX_WORKERS) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
  emit("download", 100, `${done} downloaded + ${skipped} cached = ${done + skipped} files`);
  return done + skipped;
}

// ── Step 3: Parquet → SQLite ──────────────────────────────────

function convertToSqlite(): void {
  const files = readdirSync(PARQUET_DIR)
    .filter((f) => f.endsWith(".parquet"))
    .sort()
    .map((f) => join(PARQUET_DIR, f));

  if (files.length === 0) {
    throw new Error(`No parquet files in ${PARQUET_DIR}`);
  }

  emit("convert", 0, `Converting ${files.length} parquets → ${basename(DB_PATH)}`);

  const db = new Database(DB_PATH);
  db.exec("PRAGMA journal_mode = OFF");
  db.exec("PRAGMA synchronous = OFF");
  db.exec("PRAGMA cache_size = -64000");
  db.exec("PRAGMA temp_store = MEMORY");
  db.exec("DROP TABLE IF EXISTS gutenberg");
  db.exec(`
    CREATE TABLE gutenberg (
      etextno INTEGER,
      book_title TEXT,
      author TEXT,
      issued TEXT,
      context TEXT
    )
  `);

  const insert = db.prepare(
    "INSERT INTO gutenberg (etextno, book_title, author, issued, context) VALUES (?, ?, ?, ?, ?)",
  );

  let totalRows = 0;
  const t0 = performance.now();

  for (let i = 0; i < files.length; i++) {
    const file = files[i];
    const parquetBytes = readFileSync(file);
    const arrowTable = readParquet(parquetBytes);
    const ipc = arrowTable.intoIPCStream();
    const table: Table = tableFromIPC(ipc);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const batchInsert = db.transaction((rows: any[][]) => {
      for (const row of rows) {
        insert.run(...row);
      }
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let batch: any[][] = [];
    for (let r = 0; r < table.numRows; r++) {
      const row = table.get(r);
      batch.push([
        Number(row.etextno),
        String(row.book_title),
        String(row.author),
        String(row.issued),
        String(row.context),
      ]);
      if (batch.length >= BATCH_SIZE) {
        batchInsert(batch);
        totalRows += batch.length;
        batch = [];
        const pct = ((i + r / table.numRows) / files.length) * 100;
        emit("convert", pct, `File ${i + 1}/${files.length} | ${totalRows.toLocaleString()} rows`);
      }
    }
    if (batch.length > 0) {
      batchInsert(batch);
      totalRows += batch.length;
    }
  }

  emit("index", 95, "Creating index...");
  db.exec("CREATE INDEX idx_etextno ON gutenberg(etextno)");
  emit("compact", 98, "Compacting (VACUUM)...");
  db.exec("VACUUM");

  const elapsed = (performance.now() - t0) / 1000;
  const count = db.query("SELECT COUNT(*) as n FROM gutenberg").get() as { n: number };
  emit("done", 100, `${count.n.toLocaleString()} rows in ${elapsed.toFixed(1)}s → ${DB_PATH}`);
  db.close();
}

// ── Main ──────────────────────────────────────────────────────

async function main() {
  emit("init", 0, "Listing parquet files...");
  const files = await listParquetFiles();
  emit("init", 5, `Found ${files.length} parquet files`);

  await downloadAll(files);
  convertToSqlite();
}

main().catch((err) => {
  emit("error", 0, String(err));
  process.exit(1);
});
